akka-actor/src/main/scala/akka/util/MessageBuffer.scala

```diff
@@ -0,0 +1,245 @@
+/**
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.util
+
+import akka.actor.ActorRef
+import akka.japi.Pair
+import akka.japi.function.Procedure2
+
+/**
+ * A non thread safe mutable message buffer that can be used to buffer messages inside actors.
+ */
+final class MessageBuffer private (
+  private var _head: MessageBuffer.Node,
+  private var _tail: MessageBuffer.Node) {
+  import MessageBuffer._
+
+  private var _size: Int = if (_head eq null) 0 else 1
+
+  /**
+   * Check if the message buffer is empty.
+   *
+   * @return if the buffer is empty
+   */
+  def isEmpty: Boolean = _head eq null
+
+  /**
+   * Check if the message buffer is not empty.
+   *
+   * @return if the buffer is not empty
+   */
+  def nonEmpty: Boolean = !isEmpty
+
+  /**
+   * How many elements are in the message buffer.
+   *
+   * @return the number of elements in the message buffer
+   */
+  def size: Int = _size
+
+  /**
+   * Add one element to the end of the message buffer.
+   *
+   * @param message the message to buffer
+   * @param ref the actor to buffer
+   * @return this message buffer
+   */
+  def append(message: Any, ref: ActorRef): MessageBuffer = {
+    val node: Node = new Node(null, message, ref)
+    if (isEmpty) {
+      _head = node
+      _tail = node
+    } else {
+      _tail.next = node
+      _tail = node
+    }
+    _size += 1
+    this
+  }
+
+  /**
+   * Remove and return the first element of the message buffer.
+   *
+   * @return the first element or an element containing null if the buffer is empty
+   */
+  def dropHead(): (Any, ActorRef) = {
+    if (nonEmpty) {
+      val node = _head
+      _head = _head.next
+      _size -= 1
+      if (isEmpty)
+        _tail = null
+      (node.message, node.ref)
+    } else {
+      (null, null)
+    }
+  }
+
+  /**
+   * Java API
+   *
+   * Remove and return the first element of the message buffer.
+   *
+   * @return the first element or an element containing null if the buffer is empty
+   */
+  def dropHeadP(): Pair[Any, ActorRef] = {
```

Comment:
ouch, that is too ugly.
What do you think of making it `def dropHead(): Unit` instead and have accessors `head(): (Any, ActorRef)` and `getHead(): Pair[Any, ActorRef]`. That is also more efficient if you don't care about the value.